packages/@aws-cdk/cdk/lib/core/tag-manager.ts

```diff
@@ -0,0 +1,225 @@
+
+    public static readonly DEFAULT_TAG_PROPS: TagProps = {
+        propagate: true,
+        propagateOverride: false,
+        overwrite: true
+    };
+
+    /**
+     * Checks if the object implements the `ITaggable` interface
+     */
+    public static isTaggable(taggable: ITaggable | any): taggable is ITaggable {
+        return ((taggable as ITaggable).tags !== undefined);
+    }
+
+    /*
+     * Internally tags will have properties set
+     */
+    private readonly _tags: FullTags = {};
+
+    /*
+     * Tags that will be reomved during `tags` method
+     */
+    private readonly blockedTags: string[] = [];
+
+    constructor(private readonly parent: Construct, initialTags: Tags  = {}) {
+        for (const key of Object.keys(initialTags)) {
+            const tag = {value: initialTags[key],
+                props: {propagate: true, propagateOverride: false}};
+            this._tags[key] = tag;
+        }
+    }
+
+    /**
+     * Converts the `tags` to a Token for use in lazy evaluation
+     */
+    public toCloudformation(): Token {
+        return new Token(() => this.toArray());
+    }
+
+    /**
+     * Adds the specified tag to the array of tags
+     *
+     * @param key The key value of the tag
+     * @param value The value value of the tag
+     * @param props A `TagProps` object for the tag @default `TagManager.DEFAULT_TAG_PROPS`
+     */
+    public setTag(key: string, value: string, tagProps: TagProps = {} ): void {
+        const props = {...TagManager.DEFAULT_TAG_PROPS, ...tagProps};
+        if (props.overwrite === false) {
+            this._tags[key] = this._tags[key] || {value, props};
+        } else {
+            this._tags[key] = {value, props};
+        }
+        const index = this.blockedTags.indexOf(key);
+        if (index > -1) {
+            this.blockedTags.splice(index, 1);
+        }
+    }
+
+    /**
+     * Removes the specified tag from the array if it exists
+     *
+     * @param key The key of the tag to remove
+     */
+    public removeTag(key: string, props: RemoveProps = {blockPropagate: true}): void {
+        if (props.blockPropagate) {
+            this.blockedTags.push(key);
+        }
+        delete this._tags[key];
+    }
+
+    /**
+     * Check that a key exists including known propagated tags
+     *
+     * This checks the tags known at the time method of invocation. If tags are
+     * added or removed after this method is invoked the result can change.
+     *
+     * @param key The key of the tag to check existence
+     */
+    public hasTag(key: string): boolean {
+        return this.tags[key] !== undefined;
+    }
+
+    /**
+     * Returns tags that meet the criteria of the filter
+     *
+     * The function operates only tags local to this contsturct. The function
+     * provides the ability to filter tags baseed on the `TagProps`.
+     *
+     * @param filter The tag props to filter tags
+     */
+    public filterTags(filter: TagProps = {}): Tags {
```

Comment:
Don't think this needs to be `public`?

Reply:
@rix0rrr -- this is called on `ancestors` in `propagatedTags()`. I think if I went back to a more pure recursive pattern I could eliminate.